Feature/rest api reordering post

ciprianimike opened the pull request:
### Description of the Change

The function `ajax_simple_page_ordering()` is the callback function of the AJAX call, it was also responsible for saving the position of the posts.

A refactor has been done, moving the sorting and posts updating part into a new `page_ordering()` function.

The REST endpoint has been added: `wp-json/simplepageordering/v1/page_ordering/`

The `rest_page_ordering()` function has been added as callback for the new REST endpoint and it calls the `page_ordering()` function.

`page_ordering()` is therefore the part of the code where the sorting of post takes place and this is now called via AJAX or REST.

### Verification Process

after taking some of the code out of the ajax_simple_page_ordering () function, I made sure that the operation remained intact, verifying the parameters sent in the request before the refactor and replicating them after the refactor.

I then sent the same parameters through the new REST endpoint and verified that the reordering is consistent, the same parameters via AJAX or REST must lead to the usual sorting results.

### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [x] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Applicable Issues

https://github.com/10up/simple-page-ordering/issues/67

jeffpaul committed:
Merge pull request #72 from 10up/update/wpversion

Bump WP tested-up-to version 5.8

ciprianimike committed:
REST API endpoint to sort pages

ciprianimike committed:
update readme and plugin version

lkraav commented on simple-page-ordering.php:
```diff
@@ -338,6 +364,67 @@ private static function check_edit_others_caps( $post_type ) {
+		 */
+		public static function rest_api_init() {
+			register_rest_route(
+				'simplepageordering/v1',
```
Is hyphenless the current REST API namespace convention?

I would expect it to mirror plugin name, like `simple-page-ordering`.

Looking at my large production site `/wp-json/` routes list, most if not all plugins do seem to register their namespaces in human readable form like `/wp-json/ithemes-security`

cadic replied:
[Naming convention](https://developer.wordpress.org/rest-api/extending-the-rest-api/adding-custom-endpoints/#namespacing) recommends to use theme/plugin slug as `vendor` name of the endpoint but there is no strict requirement to use hyphenated or CamelCase.

In this case, since it's the first REST endpoint in this plugin, would be nice to have consistent `simple-page-ordering` vendor name in the route, same as plugin slug and textdomain.

cadic commented on simple-page-ordering.php:
```diff
@@ -338,6 +364,67 @@ private static function check_edit_others_caps( $post_type ) {
+
+			// check and make sure we have what we need
+			if ( empty( $post_id ) || ( ! isset( $previd ) && ! isset( $nextid ) ) ) {
+				wp_send_json_error();
```
REST Callback [expected to return the result](https://developer.wordpress.org/rest-api/extending-the-rest-api/adding-custom-endpoints/#return-value) of the action instead of error processing or output inside the callback function. If an error happens, we need to return `new WP_Error( ... )`
> Importantly, a REST API route’s callback should always return data; it shouldn’t attempt to send the response body itself.

cadic commented on simple-page-ordering.php:
```diff
@@ -338,6 +364,67 @@ private static function check_edit_others_caps( $post_type ) {
+
+			self::page_ordering( $post_id, $previd, $nextid, $start, $excluded );
+
+			wp_send_json_error();
```
After successful processing, REST endpoint callback expected to return [WP_REST_Response](https://developer.wordpress.org/reference/classes/wp_rest_response/)

cadic commented on simple-page-ordering.php:
```diff
@@ -141,17 +151,30 @@ public static function ajax_simple_page_ordering() {
+			$previd   = empty( $previd ) ? false : (int) $previd;
+			$nextid   = empty( $nextid ) ? false : (int) $nextid;
+			$start    = empty( $start ) ? 1 : (int) $start;
+			$excluded = empty( $excluded ) ? array( $post_id ) : array_filter( (array) json_decode( $excluded ), 'intval' ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
```
Is this second validation really necessary after validating $_POST arguments in `ajax_simple_page_ordering()`?

cadic commented on simple-page-ordering.php:
```diff
@@ -338,6 +364,67 @@ private static function check_edit_others_caps( $post_type ) {
+			$previd   = empty( $_GET['previd'] ) ? false : (int) $_GET['previd'];
+			$nextid   = empty( $_GET['nextid'] ) ? false : (int) $_GET['nextid'];
+			$start    = empty( $_GET['start'] ) ? 1 : (int) $_GET['start'];
+			$excluded = empty( $_GET['excluded'] ) ? array( $_GET['id'] ) : array_filter( (array) json_decode( $_GET['excluded'] ), 'intval' ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
```
[REST arguments](https://developer.wordpress.org/rest-api/extending-the-rest-api/adding-custom-endpoints/#arguments) are passed to the callback as a single argument `rest_page_ordering(  WP_REST_Request $request )` and should be accessed as incoming array values, not from $_GET or $_POST.

lkraav replied:
Also, many PHPCS rulesets advise using `filter_input()` instead of accessing request globals directly.

ciprianimike committed:
fix rest api naming convention

ciprianimike committed:
simple-page-ordering refactor

- Added PHPCS via composer
- Added 10UP PHPCS ruleset
- Clean up
- page_ordering returns object or WP_Error
- added REST responses

cadic commented on readme.txt:
```diff
@@ -3,7 +3,7 @@ Contributors: 10up, jakemgold, welcher, helen, thinkoomph
 Tags: order, re-order, ordering, pages, page, manage, menu_order, hierarchical, ajax, drag-and-drop, admin
 Requires at least: 3.8
-Tested up to: 5.8
+Tested up to: 5.7
```
Do we really downshift this from 5.8? Also includes README.md, CREDITS.md and CHANGELOG.md

ciprianimike replied:
merged `develop` into the feature branch, resolved conflicts and added documentation

cadic commented on simple-page-ordering.php:
```diff
@@ -141,26 +168,54 @@ public static function ajax_simple_page_ordering() {
+		 * @param int    $previd   The previous post ID.
+		 * @param int    $nextid   The next post ID.
+		 * @param int    $start    The start index.
+		 * @param string $excluded Array of post IDs.
```
Doc type for `$excluded` seems to be `array`

cadic commented on simple-page-ordering.php:
```diff
@@ -338,6 +407,79 @@ private static function check_edit_others_caps( $post_type ) {
+		 * @param WP_REST_Request $request The REST request object.
+		 */
+		public static function rest_page_ordering( WP_REST_Request $request ) {
+			$post_id  = empty( $request->get_param( 'id' ) ) ? false : (int) $request->get_param( 'id' );
```
To validate, sanitize, specify mandatory arguments and default values we could use `register_rest_route()` at `Simple_Page_Ordering::rest_api_init()` instead of manual handle here.

Please refer to the [REST handbook](https://developer.wordpress.org/rest-api/extending-the-rest-api/adding-custom-endpoints/#:~:text=Arguments%20are%20defined%20as%20a%20map%20in%20the%20key%20args%20for%20each%20endpoint)

ciprianimike replied:
we are already using it at L#422
probably this checks are redundant but consistent with the ones done in ajax callback L#145

cadic replied:
@ciprianimike at L#422 we only have `description` and `type` for each argument. I mean we could add `required` and `default`. It will help to return detailed error REST responses natively (without manual return like L#462)

cadic reviewed (changes requested):
@ciprianimike thank you for the update. There are couple minor changes left and maybe a merge conflict (version bumb and readme files)

cadic commented on simple-page-ordering.php:
```diff
@@ -141,26 +168,54 @@ public static function ajax_simple_page_ordering() {
+				die( -1 );
+			}
+
+			die( wp_json_encode( $result ) );
```
This is not related to the issue itself, but using `die()` to return AJAX errors and success is obsolete, better to go with `wp_send_json_error()` and `wp_send_json_success()`

ciprianimike replied:
I guess it should go to a new issue as it is a practice used elsewhere in this function and we need to check another section of the plugin that has nothing to do with the REST functionality

ciprianimike committed:
Merge remote-tracking branch 'origin/develop' into feature/REST-API-reordering-post

ciprianimike committed:
v.2.3.5 documentation

ciprianimike committed:
fix tested v5.8

ciprianimike committed:
update page_ordering doc

ciprianimike committed:
removed screen_id for REST API, used only in the admin ajax callback

ciprianimike committed:
add REST register_rest_route args parameters `default` and `required`

ciprianimike commented:
@cadic got it, added parameters `required` and `default`

cadic reviewed (approved):
Thanks @ciprianimike, it's a great new feature!
